sdk/src/types/block/block_id.rs

```diff
@@ -1,11 +1,112 @@
 // Copyright 2020-2021 IOTA Stiftung
 // SPDX-License-Identifier: Apache-2.0
 
-impl_id!(
-    pub BlockId,
-    32,
-    "A block identifier, the BLAKE2b-256 hash of the block bytes. See <https://www.blake2.net/> for more information."
-);
+use super::{slot::SlotIndex, ConvertTo};
+use crate::types::block::Error;
 
+impl_id!(pub BlockHash, 32, "The hash of a [`Block`].");
+
+impl BlockHash {
+    #[cfg(target_endian = "little")]
+    pub fn with_slot_index(self, slot_index: SlotIndex) -> BlockId {
+        BlockId { hash: self, slot_index }
+    }
+
+    #[cfg(target_endian = "big")]
+    pub fn with_slot_index(self, slot_index: SlotIndex) -> BlockId {
+        BlockId {
+            hash: self,
+            slot_index: slot_index.to_le().into(),
+        }
+    }
+}
+
+/// A block identifier.
+#[derive(Clone, Copy, Eq, PartialEq, Hash, Ord, PartialOrd, Debug, packable::Packable)]
+#[packable(unpack_error = Error)]
+#[repr(C)]
+pub struct BlockId {
+    pub(crate) hash: BlockHash,
+    // INPORTANT: On big-endian systems this value is misrepresented because it is transmuted directly
+    // from bytes, so the getter below handles that conversion. Do not access it directly.
+    slot_index: SlotIndex,
+}
+
+impl BlockId {
+    /// The length of a [`BlockId`]
+    pub const LENGTH: usize = 40;
+
+    pub fn new(bytes: [u8; Self::LENGTH]) -> Self {
+        unsafe { core::mem::transmute(bytes) }
+    }
+
+    /// Returns the [`BlockId`]'s slot index part.
+    #[cfg(target_endian = "little")]
+    pub fn slot_index(&self) -> SlotIndex {
+        self.slot_index
+    }
+
+    /// Returns the [`BlockId`]'s slot index part.
+    #[cfg(target_endian = "big")]
+    pub fn slot_index(&self) -> SlotIndex {
+        self.slot_index.to_le().into()
+    }
+}
+
+impl AsRef<[u8]> for BlockId {
+    fn as_ref(&self) -> &[u8] {
+        unsafe { core::mem::transmute::<_, &[u8; Self::LENGTH]>(self) }
```

Comment:
so this would technically go wrong on be systems.
maybe set a feature flag warn when its called?

Reply:
no, this is correct on big or little endian because the underlying memory is always the same